### PR TITLE
feat: support monitoring WebSocket endpoints

### PR DESCRIPTION
**Pull request by heitorPB**
WebSocket endpoints are automatically identified by the URL protocol specification: `wss://` or `ws://`. The request body is used as the "message" written to the server, and the answer is stored in the `[BODY]`.

Optionally, the user can set the `jsonrpc` flag to automatically wrap the request body in a JSON RPC 2.0 method call.

### CODE REVIEW

File: core/endpoint.go [hunk 1]
```diff
@@ -97,6 +98,9 @@ type Endpoint struct {
 	// GraphQL is whether to wrap the body in a query param ({"query":"$body"})
 	GraphQL bool `yaml:"graphql,omitempty"`
 
+	// JsonRPC is wether to wrap the body in as a JSON RPC 2.0 method call. First word becomes method name and the rest becomes parameters
```
Comment (0xAleksaOpacic) [inline core/endpoint.go]: Small typo: wether->whether

File: core/websocket.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+	if err != nil {
+		result.AddError("Error dialing WS:" + err.Error())
+		return
+	}
+	result.Connected = true
+
+	// Write message
+	if _, err := ws.Write([]byte(message)); err != nil {
+		//fmt.Println("Error writing:", err)
+		result.AddError("Erro writing WS message" + err.Error())
```
Comment (0xAleksaOpacic) [inline core/websocket.go]: Typo Erro -> Error, also in other error

File: core/websocket.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+	ws, err := websocket.DialConfig(config)
+	if err != nil {
+		result.AddError("Error dialing WS:" + err.Error())
+		return
+	}
+	result.Connected = true
+
+	// Write message
+	if _, err := ws.Write([]byte(message)); err != nil {
+		//fmt.Println("Error writing:", err)
```
Comment (0xAleksaOpacic) [inline core/websocket.go]: Leftover comment we can remove it

File: core/websocket.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+		result.AddError("Error configuring WS connection:" + err.Error())
+		return
+	}
+
+	// Dial URL
+	ws, err := websocket.DialConfig(config)
+	if err != nil {
+		result.AddError("Error dialing WS:" + err.Error())
+		return
+	}
```
Comment (0xAleksaOpacic) [inline core/websocket.go:27]: When we create ws connection we should make sure to remove it once we finish using it
`defer ws.Close()`

File: core/websocket.go [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+	// Write message
+	if _, err := ws.Write([]byte(message)); err != nil {
+		//fmt.Println("Error writing:", err)
+		result.AddError("Erro writing WS message" + err.Error())
+		return
+	}
+
+	// Read message (1 kByte)
+	var msg = make([]byte, 1024)
```
Comment (0xAleksaOpacic) [inline core/websocket.go]: suggestion, whenever you have these magic numbers, put them in the const above method
```
const (
	Origin         = "http://localhost/"
	MaxMessageSize = 1024
)
```

File: core/endpoint.go [hunk 6]
```diff
@@ -189,6 +195,21 @@ func (endpoint *Endpoint) ValidateAndSetDefaults() error {
...
+
+		if len(endpoint.Body) > 0 {
+			slices := strings.Split(endpoint.Body, " ")
+			method_name = slices[0]
+
+			if len(slices) > 1 {
+				method_parameters = strings.Join(slices[1:], ",")
+			}
+		}
+		endpoint.Body = "{\"jsonrpc\": \"2.0\", \"id\": 1, \"method\": \"" + method_name + "\", \"parameters\": [" + method_parameters + "]}\n"
```
Comment (0xAleksaOpacic) [inline core/endpoint.go:211]: If this condition is not met
`if len(endpoint.Body) > 0 {`
if someone does not specify the body correctly we can have an incorrect endpoint. body. Maybe we can handle this case as error

**Review by 0xAleksaOpacic — APPROVED** [review]
Thank you for contributing ad adding this AWESOME feature 🚀💪
I left some minor comments, nothing is critical. The code seems great and also thank you for adding this in the readme as well 😎

I haven't tested this code manually, please ping me if i should

File: core/endpoint.go (hunk 6 above)
Comment (heitorPB) [reply core/endpoint.go:211]: I thought about that. But we can send empty messages via WebSockets:

```bash
$ echo "" | websocat --jsonrpc wss://arbitrum-caddi.lavapro.xyz/one/ws
{"jsonrpc":"2.0","id":1,"error":{"code":-32600,"message":"invalid request"}}
```

What kind of error would you suggest here?

File: core/websocket.go (hunk 4 above)
Comment (heitorPB) [reply core/websocket.go:27]: I thought Go's GC would take care of that?

File: core/websocket.go (hunk 5 above)
Comment (heitorPB) [reply core/websocket.go]: Thanks!

**Pull request merged**